### ISSUE TITLE
Update `AccessorProp` to throw error for default value

### ISSUE THREAD

**Issue by JordanMartinez** [purescm/purescm#113]
See for context https://github.com/purescm/purescm/pull/112/files#r1157034191

[hashtable-ref](https://scheme.com/tspl4/objects.html#./objects:s281)

**Comment by purefunctor** [comment]
My assumption was that missing values would be caught during type checking, unless some coercion was going on. I reckon we can define a runtime function for this though—basically a throw-if-not-exists check or something along those lines.

**Comment by JordanMartinez** [comment]
Since we have `unsafeCoerce`, I think we should assume that someone (e.g. me) will abuse this at some point. Otherwise the below code actually works and could lead to hard-to-debug issues:
```
foo :: forall a. (a -> Boolean) -> Array a -> Array a
foo findIt = Array.findBy findIt

baz = foo (\x -> (unsafeCoerce x).bar) [ 1, 2, 3, 4 ]
```

**Comment by f-f** [comment]
We had this happen in the previous iteration of purescm, where the codegen was not correct and the thing was blowing up in an unspecified location - it took way too long to debug.
I'd say an error should be thrown there, and it should be unique (e.g. by attaching a random number or the source location)

**Comment by turlando** [comment]
Instead of hashtable-ref, we need something like

```scheme
(cond
  [(hashtable-contains? ht key) (hashtable-ref ht key '())]
  [else (error ...)])
```

The dirty solution is to emit this code at each record lookup. The clean solution would be to include a macro in the runtime.

**Comment by JordanMartinez** [comment]
Does `hashtable-ref ht key (throw-error)` not work? Otherwise, it seems like the above code will compute the hash for the key twice.

**Comment by JordanMartinez** [comment]
Ah, thanks for clarifying.

**Comment by turlando** [comment]
Yes, thanks @f-f for expliciting my cryptic message. Basically it is a consequence of the Scheme's applicative evaluation order. In order to delay the evaluation of `error` we need a special form (an expression that has special evaluation rules, such as let, cond, lambda, define, quote, ...).

Another consideration is that doing a `hashtable-contains?` and then a `hashtable-ref` is definitely not thread safe. But in that regard, I'm not even sure if `hashtable-ref` is thread safe per se.

> Otherwise, it seems like the above code will compute the hash for the key twice.

That is a good observation. Maybe we can write our own procedure/macro in function of the intrinsics that hashtable-ref internally uses (if they are exposed).

**Comment by turlando** [comment]
I made a proof of concept (#117) addressing this issue.

**Comment by JordanMartinez** [comment]
Closed by #138
